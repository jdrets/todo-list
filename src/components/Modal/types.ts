export interface ModalTypes {
  title: string,
  open: boolean,
  onClose: () => void,
}
