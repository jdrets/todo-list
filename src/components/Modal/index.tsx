import React, { FunctionComponent, useEffect } from 'react'
import { Close as CloseIcon } from 'react-ionicons'

import {
  Wrapper,
  Container,
  Header,
  Title,
  CloseIconWrapper,
  Backdrop
} from './styles'
import { ModalTypes } from './types'

const Modal: FunctionComponent<ModalTypes> = ({
  children,
  title,
  open,
  onClose
}) => {
  useEffect(() => {
    document.addEventListener('keydown', (event: KeyboardEvent) => {
      if (event.key === 'Escape' && open) {
        onClose()
      }
    })
  }, [open])

  return (
    open && <Wrapper>
      <Backdrop onClick={onClose} />
      <Container>
        <Header>
          <Title>{title}</Title>
          <CloseIconWrapper onClick={onClose} data-testid="modal-close-button">
            <CloseIcon />
          </CloseIconWrapper>
        </Header>
        {children}
      </Container>
    </Wrapper>
  )
}

export default Modal
