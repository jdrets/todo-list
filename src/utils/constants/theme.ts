const palette = {
  purpleDark: '#4B00BC',
  purple: '#6600FF',
  purpleMedium: '#8533FF',
  purpleLight: '#E3D0FF',
  green: '#59BC6F',
  greenLight: '#E6F5E2',
  red: '#EB5757',
  redLight: '#FCE4E4',
  yellow: '#DD9900',
  yellowLight: '#FEF1D8',
  white: '#FFFFFF',
  black: '#000000',
  gray: '#898989',
  grayLight: '#E0E0E0'
}

const theme = {
  colors: {
    primaryColor: palette.purple,
    primaryColorMedium: palette.purpleMedium,
    primaryColorLight: palette.purpleLight,
    primaryColorDark: palette.purpleDark,
    ...palette
  },
  layout: {
    breakpoints: {
      medium: '768px'
    },
    mobile: {
      width: '100%',
      padding: '0 24px'
    },
    desktop: {
      width: '1050px'
    }
  }
}

export default theme
