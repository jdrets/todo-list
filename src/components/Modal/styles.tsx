import styled from 'styled-components'

import theme from '../../utils/constants/theme'

export const Wrapper = styled.div<any>`
  align-items: center;
  display: flex;
  height: 100vh;
  justify-content: center;
  left: 0;
  position: fixed;
  top: 0;
  width: 100%;
  z-index: 3;
`

export const Backdrop = styled.div<any>`
  background-color: rgba(0, 0, 0, 0.2);
  height: 100vh;
  position: absolute;
  width: 100%;
`

export const Container = styled.div`
  background-color: white;
  min-height: 100vh;
  padding: 24px;
  position: relative;
  width: 100%;
  
  @media (min-width: ${theme.layout.breakpoints.medium}) {
    border-radius: 12px;
    height: auto;
    min-width: 350px;
    max-width: 90%;
    min-height: auto;
    width: auto;
  }
`

export const Header = styled.div`
  align-items: center;
  display: flex;
  justify-content: space-between;
  padding-bottom: 24px;
`

export const Title = styled.h2`
  color: ${theme.colors.primaryColor};
  font-size: 24px;
  padding-right: 24px;
  word-break: break-word;
`

export const CloseIconWrapper = styled.button`
  background-color: transparent;
  border: none;
  cursor: pointer;
  
  svg {
    align-items: center;
    display: flex;
    fill: ${theme.colors.gray};
    height: 40px;
    width: 40px;
  }
`
