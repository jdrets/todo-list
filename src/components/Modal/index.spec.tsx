import React from 'react'
import { ThemeProvider } from 'styled-components'

import { render, fireEvent } from '@testing-library/react'

import Modal from '.'
import theme from '../../utils/constants/theme'

const setup = (mockData = {}) => {
  const mock = {
    onClose: jest.fn(),
    title: 'dummy title',
    open: true
  }
  return render(
    <ThemeProvider theme={theme}>
      <Modal {...mock} {...mockData}>
        Modal content
      </Modal>
    </ThemeProvider>
  )
}

describe('<Modal />', () => {
  test('should match with snapshot', () => {
    const { container } = setup()

    expect(container).toMatchSnapshot()
  })

  test('should close modal when press Escape key and modal is open', () => {
    const onCloseMock = jest.fn()
    setup({
      open: true,
      onClose: onCloseMock
    })

    fireEvent.keyDown(document, {
      key: 'Escape'
    })

    expect(onCloseMock).toHaveBeenCalled()
  })

  test('should not close modal when press Escape key and modal is not open', () => {
    const onCloseMock = jest.fn()
    setup({
      open: false,
      onClose: onCloseMock
    })

    fireEvent.keyDown(document, {
      key: 'Escape'
    })

    expect(onCloseMock).toHaveBeenCalledTimes(0)
  })

  test('should not render when open prop is false', () => {
    const mockData = {
      open: false
    }
    setup(mockData)
    const button = document.getElementsByTagName('button')[0]

    expect(button).toBeFalsy()
  })

  test('should call onClose function when close button is clicked', () => {
    const mockOnClose = jest.fn()
    const mockData = {
      onClose: mockOnClose
    }
    const { getByTestId } = setup(mockData)
    const button = getByTestId('modal-close-button')

    fireEvent.click(button)

    expect(mockOnClose).toHaveBeenCalled()
  })
})
