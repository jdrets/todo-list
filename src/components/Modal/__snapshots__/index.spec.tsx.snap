// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Modal /> should match with snapshot 1`] = `
<div>
  <div
    class="sc-gtssRu fYrUmT"
  >
    <div
      class="sc-dlnjPT eBfcOO"
    />
    <div
      class="sc-hKFyIo fwUtbv"
    >
      <div
        class="sc-eCApGN hzcBsN"
      >
        <h2
          class="sc-jSFkmK kvDxJQ"
        >
          dummy title
        </h2>
        <button
          class="sc-gKAblj doETKQ"
          data-testid="modal-close-button"
        >
          <span
            class="sc-bdnylx kpuqbH"
            color="#000000"
            height="22px"
            rotate="0"
            width="22px"
          >
            <svg
              class=""
              viewBox="0 0 512 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              
              <path
                d="M289.94 256l95-95A24 24 0 00351 127l-95 95-95-95a24 24 0 00-34 34l95 95-95 95a24 24 0 1034 34l95-95 95 95a24 24 0 0034-34z"
              />
            </svg>
          </span>
        </button>
      </div>
      Modal content
    </div>
  </div>
</div>
`;
